File: 0029-divide-two-integers/0029-divide-two-integers.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        # 2147483647
        d, dv, output = abs(dividend), abs(divisor), 0
        while d >= dv:
            temp, mul = dv, 1
            while d >= temp:
                d -= temp
                output += mul
                mul += mul
                temp += temp
        if dividend < 0 and divisor >= 0 or divisor < 0 and dividend >= 0:
            output = -output
        return min(2147483647, max(-2147483648, output))

File: 0029-divide-two-integers/0029-divide-two-integers.go
func divide(dividend int, divisor int) int {
  d, dv := 0, 0
  if dividend >= 0 {
    d = dividend
  } else {
    d = -dividend
  }
  if divisor >= 0 {
    dv = divisor
  } else {
    dv = -divisor
  }
  output, temp, mul := 0, 0, 0
  for d >= dv {
    temp, mul = dv, 1
    for d >= temp {
      d, output, mul, temp = d-temp, output+mul, mul+mul, temp+temp
    }
  }
  if dividend < 0 && divisor >= 0 || divisor < 0 && dividend >= 0 {
    output = -output
  }
  return min(2147483647, max(-2147483648, output))
}